Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -0,0 +1,448 @@
+  rw [weierstrassDiv, dif_neg (by simp)]
+
+@[simp]
+theorem weierstrassMod_zero_right [IsPrecomplete (IsLocalRing.maximalIdeal A) A] : f %ʷ 0 = 0 := by
+  rw [weierstrassMod, dif_neg (by simp)]
+
+theorem degree_weierstrassMod_lt [IsPrecomplete (IsLocalRing.maximalIdeal A) A] :
+    (f %ʷ g).degree < (g.map (IsLocalRing.residue A)).order.toNat := by
+  rw [weierstrassMod]
+  split_ifs with hg
+  · exact degree_trunc_lt _ _
+  · nontriviality A
+    rw [Polynomial.degree_zero]
+    exact WithBot.bot_lt_coe _
+
+section
+
+variable {g} (hg : g.map (IsLocalRing.residue A) ≠ 0)
+include hg
+
+theorem isWeierstrassDivision_weierstrassDiv_weierstrassMod
+    [IsAdicComplete (IsLocalRing.maximalIdeal A) A] :
+    f.IsWeierstrassDivision g (f /ʷ g) (f %ʷ g) := by
+  simp_rw [weierstrassDiv, weierstrassMod, dif_pos hg]
+  exact (IsWeierstrassDivisor.of_map_ne_zero hg).isWeierstrassDivisionAt_div_mod f
+
+theorem eq_mul_weierstrassDiv_add_weierstrassMod
+    [IsAdicComplete (IsLocalRing.maximalIdeal A) A] :
+    f = g * (f /ʷ g) + (f %ʷ g) := by
+  simp_rw [weierstrassDiv, weierstrassMod, dif_pos hg]
+  exact ((IsWeierstrassDivisor.of_map_ne_zero hg).isWeierstrassDivisionAt_div_mod f).2
+
+/-- The quotient `q` and remainder `r` in the Weierstrass division `0 / g` are equal to `0`. -/
+theorem eq_zero_of_weierstrass_division [IsHausdorff (IsLocalRing.maximalIdeal A) A]
+    {q : A⟦X⟧} {r : A[X]}
+    (hr : r.degree < (g.map (IsLocalRing.residue A)).order.toNat)
+    (heq : g * q = r) : q = 0 ∧ r = 0 :=
+  (IsWeierstrassDivisor.of_map_ne_zero hg).eq_zero_of_mul_eq hr heq
+
+/-- The quotient `q` and remainder `r` in the Weierstrass division are unique. -/
+theorem eq_of_weierstrass_division [IsHausdorff (IsLocalRing.maximalIdeal A) A]
+    {q q' : A⟦X⟧} {r r' : A[X]}
+    (hr : r.degree < (g.map (IsLocalRing.residue A)).order.toNat)
+    (hr' : r'.degree < (g.map (IsLocalRing.residue A)).order.toNat)
+    (heq : g * q + r = g * q' + r') : q = q' ∧ r = r' :=
+  (IsWeierstrassDivisor.of_map_ne_zero hg).eq_of_mul_add_eq_mul_add hr hr' heq
+
+variable {f} in
+/-- The quotient `q` and remainder `r` in the Weierstrass division are unique. -/
+theorem IsWeierstrassDivision.elim [IsHausdorff (IsLocalRing.maximalIdeal A) A]
+    {q q' : A⟦X⟧} {r r' : A[X]}
+    (H : f.IsWeierstrassDivision g q r) (H2 : f.IsWeierstrassDivision g q' r') : q = q' ∧ r = r' :=
+  g.eq_of_weierstrass_division hg H.1 H2.1 (H.2.symm.trans H2.2)
+
+/-- The quotient `q` and remainder `r` in the Weierstrass division `0 / g` are equal to `0`. -/
+theorem IsWeierstrassDivision.eq_zero [IsHausdorff (IsLocalRing.maximalIdeal A) A]
+    {q : A⟦X⟧} {r : A[X]}
+    (H : IsWeierstrassDivision 0 g q r) : q = 0 ∧ r = 0 :=
+  H.elim hg (g.isWeierstrassDivisionAt_zero _)
+
+variable {f} in
+/-- The quotient `q` and remainder `r` in the Weierstrass division `f / g` are unique,
+in particular, equal to `f /ʷ g` and `f %ʷ g`. -/
+theorem IsWeierstrassDivision.eq_weierstrassDiv_weierstrassMod
```

Comment:
I still don't get the point of having `eq_of_weierstrass_division` and `IsWeierstrassDivision.elim`: of course they're needed to prove this theorem, that is all we want. What about inlining the two previous result in this proof?

Reply:
Ah I didn't notice it. Let's keep it then (can you add a sentence to the docstring of `IsWeierstrassDivision.elim` saying that in the complete case one probably wants to use this result?).

And yes, please remove `eq_XXX_of_XXX`.

Reply:
Done.